### DISCUSSION TITLE
Macros

### DISCUSSION THREAD

**Discussion by MystPi** [DanielXMoore/Civet#449] [Ideas]
I know that this would be a huge undertaking, but I think having macros in Civet would make the language really incredible. Not only that, but it would also be a great selling point for the language as it would add a new feature that most compile-to-js languages don't have.

When it comes to design, I immediately think of the unmaintained [Sweet.js](https://www.sweetjs.org/) compiler. There currently exists nothing like it, and having something like it in the language itself would be amazing. Of course, the macro system could be completely different; I just love the raw power Sweet.js gives you. Here's an idea:

```civet
syntax hi(ctx)
  %"console.log('Hello, world!')"

hi
# Compiles to:
# console.log('Hello, world!')

syntax new(ctx)
  ident := ctx.next().value
  params := ctx.next().value
  %"#{ident}.create #{params}"

new Test 1234
# Compiles to:
# Test.create(1234)

syntax let(ctx)
  ident := ctx.next().value
  ctx.next()
  init := ctx.expand('expr').value # Or whatever rule Civet's parser uses for expressions
  %"""
    ((#{ident}) =>
      #{ctx}
    )(#{init})
  """

let name = "MystPi"
console.log name
# Compiles to:
# ((name) => {
#   console.log(name);
# })("MystPi")

# More examples at https://www.sweetjs.org/doc/tutorial
```
This is just something to think about; I'm not sure how feasible it would be to actually implement. I've just always wanted a macro system in some form of JavaScript.

**Comment by shurko0x4cfd** [comment]
The most elegant ways to extend the language for new features that I know is the Forth language mechanisms. I would like something like that.
